File: CBLClient/Apps/CBLTestServer-Android/app/src/main/java/com/couchbase/CouchbaseLiteServ/server/RequestHandler/ConflictRequestHandler.java
package com.couchbase.CouchbaseLiteServ.server.RequestHandler;


import com.couchbase.CouchbaseLiteServ.server.Args;
import com.couchbase.lite.Conflict;
import com.couchbase.lite.ConflictResolver;
import com.couchbase.lite.Document;

public class ConflictRequestHandler {

    public ConflictResolver resolver(Args args) {
        String conflictType = args.get("conflict_type");
        if (conflictType.equals("mine")) {
            return new ReplicationMine();
        } else if (conflictType.equals("theirs")) {
            return new ReplicationTheirs();
        } else if (conflictType.equals("base")) {
            return new ReplicationBase();
        } else {
            return null;
        }
    }

    public Document getBase(Args args){
        Conflict conflict = args.get("conflict");
        return conflict.getBase();
    }

    public Document getMine(Args args){
        Conflict conflict = args.get("conflict");
        return conflict.getMine();
    }

    public Document getTheirs(Args args){
        Conflict conflict = args.get("conflict");
        return conflict.getTheirs();
    }
}

class ReplicationMine implements ConflictResolver {
    @Override
    public Document resolve(Conflict conflict) {
        return conflict.getMine();
    }
}

class ReplicationTheirs implements ConflictResolver {
    @Override
    public Document resolve(Conflict conflict) {
        return conflict.getTheirs();
    }
}

class ReplicationBase implements ConflictResolver {
    @Override
    public Document resolve(Conflict conflict) {
        return conflict.getBase();
    }
}

class GiveUp implements ConflictResolver {
    @Override
    public Document resolve(Conflict conflict) {
        return null;
    }
}

File: CBLClient/Apps/CBLTestServer-iOS/CBLTestServer-iOS/Server/ConflictRequestHandler.swift
//
//  ConflictRequestHandler.swift
//  CBLTestServer-iOS
//
//  Created by Raghu Sarangapani on 12/26/17.
//  Copyright © 2017 Raghu Sarangapani. All rights reserved.
//

import Foundation
import CouchbaseLiteSwift


public class ConflictRequestHandler {
    public static let VOID = NSObject()
    fileprivate var _pushPullReplListener:NSObjectProtocol?
    
    public func handleRequest(method: String, args: Args) throws -> Any? {
        switch method {
        //////////////
        // Conflict //
        //////////////
        case "conflict_resolver":
            let conflict_type: String? = args.get(name: "conflict_type")
            if conflict_type == "mine" {
                return ReplicationMine()
                
            } else if conflict_type == "theirs" {
                return ReplicationTheirs()
            }
            else if conflict_type == "base" {
                return ReplicationBase()
            }
            else {
                func resolve(conflict: Conflict) -> Document? {
                    return nil
                }
            }

        default:
            throw RequestHandlerError.MethodNotFound(method)
        }
        return ConflictRequestHandler.VOID
    }
}

class ReplicationMine: ConflictResolver {
    func resolve(conflict: Conflict) -> Document? {
        return conflict.mine
    }
}

class ReplicationTheirs: ConflictResolver {
    func resolve(conflict: Conflict) -> Document? {
        return conflict.theirs
    }
}

class ReplicationBase: ConflictResolver {
    func resolve(conflict: Conflict) -> Document? {
        return conflict.base
    }
}

class GiveUp: ConflictResolver {
    func resolve(conflict: Conflict) -> Document? {
        return nil
    }
}


